cmd/util/flags/cliflags/job.go

```diff
@@ -0,0 +1,232 @@
+package cliflags
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/bacalhau-project/bacalhau/cmd/util/flags"
+	"github.com/bacalhau-project/bacalhau/cmd/util/parse"
+	"github.com/bacalhau-project/bacalhau/pkg/model"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+)
+
+type JobSettings struct {
+	name        string
+	namespace   string
+	jobType     string
+	priority    int
+	count       int
+	constraints string
+	labels      map[string]string
+
+	// TODO(forrest): remove these fields and their usage when we complete deprecation of legacy flag names.
+	// tracked via https://github.com/bacalhau-project/bacalhau/issues/3838
+	legacy *LegacyJobFlags
+	// we hold a reference to the command so we may check if users provided legacy or new flag names.
+	cmd *cobra.Command
+}
+
+func (j *JobSettings) Name() string {
+	return j.name
+}
+
+func (j *JobSettings) Namespace() string {
+	return j.namespace
+}
+
+func (j *JobSettings) Type() string {
+	if j.cmd.Flags().Changed("target") {
+		jobType := models.JobTypeBatch
+		if j.legacy.targetingMode == model.TargetAll {
+			jobType = models.JobTypeOps
+		}
+		return jobType
+	}
+	return j.jobType
+}
+
+func (j *JobSettings) Priority() int {
+	return j.priority
+}
+
+func (j *JobSettings) Count() int {
+	if j.cmd.Flags().Changed("concurrency") {
+		return j.legacy.concurrency
+	}
+	return j.count
+}
+
+func (j *JobSettings) Constraints() ([]*models.LabelSelectorRequirement, error) {
+	if j.cmd.Flags().Changed("selector") {
+		req, err := parse.NodeSelector(j.legacy.selectors)
+		if err != nil {
+			return nil, err
+		}
+		out := make([]*models.LabelSelectorRequirement, 0, len(req))
+		for _, c := range req {
+			tmp := models.LabelSelectorRequirement(c)
+			out = append(out, &tmp)
+		}
+		return out, nil
+	}
+	return parse.NodeSelectorV2(j.constraints)
+}
+
+const RegexString = "A-Za-z0-9._~!:@,;+-"
+
+func IsSafeAnnotation(s string) bool {
+	matches := SafeAnnotationRegex().FindString(s)
+	return matches == ""
+}
+
+func returnAllEmojiString() string {
+	//nolint:lll,stylecheck // exception for one time string
```

Comment:
is this the best way to do this? we are using label selector library from k8s, and I am sure we can use something from that library for label validation

Reply:
This code is ported over from the [legacy job CLI package](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/legacyjob/emojiarray.go) in an attempt to preserve backwards compatibility. I think the better way would be using the validation provided by the K8s selectors, but that would mean a breaking change to the `labels` flag. 

Generally speaking there isn't a straight forward way maintain compatibility across this flag since
- Previously labels were a string slice: https://github.com/bacalhau-project/bacalhau/blob/main/pkg/model/job.go#L230
- Now they are a map of strings: https://github.com/bacalhau-project/bacalhau/blob/main/pkg/models/job.go#L99

To accommodate this change we must accept and parse the flags differently on the CLI and include a [plural flag name](https://github.com/bacalhau-project/bacalhau/pull/4020#discussion_r1615175328) for the map case. When we accept a slice instead of a map we migrate it to a map with empty keys: https://github.com/bacalhau-project/bacalhau/blob/main/pkg/models/migration/legacy/from.go#L34. But empty keys are considered invalid by the k8s label package.

My suggestion here, if you're willing to move forward with it, would be to drop the flag that accepts a slice and make this a breaking change only in the context of this flag. what do you think?

Reply:
We don't need separate flags. What I am trying to say is if k8s selectors enable emoji's, then we don't need to prevent them, and if they have additional validations, then we should enforce them. We shouldn't categorize this as a breaking change as most likely the additional validations are not common cases that will break our limited set of existing users.

As for existing labels being a slice, and new ones being key/value, we've discussed this before that we can still accept a slice of strings and parse them as keys with empty values. We don't need to document this so we can stop the bleeding, but we don't need to break all existing submissions and we don't need to continue supporting two different flags

Reply:
From our sync on this: We will maintain one flag for labels that accepts either keys or key values and leverage the k8s validation.